[spark-sql] a better exception message than "scala.MatchError" for unsupported types in Schema creation

eranation opened the pull request:
Currently if trying to register an RDD (or DataFrame in 1.3) as a table that has types that have no supported Schema representation (e.g. type "Any") - it would throw a match error. e.g. scala.MatchError: Any (of class scala.reflect.internal.Types$ClassNoArgsTypeRef)

This fix is just to have a nicer error message than a MatchError

eranation committed:
better exception than scala.MatchError: Any 

Currently if trying to register an RDD as a table that has types that have no supported Schema representation - it would throw a match error. e.g. scala.MatchError: Any (of class scala.reflect.internal.Types$ClassNoArgsTypeRef)

This fix is just to have a nicer error message thank a MatchError

AmplabJenkins commented:
Can one of the admins verify this patch?

marmbrus commented:
ok to test

SparkQA commented:
[Test build #29332 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29332/consoleFull) for   PR 5235 at commit [`524be86`](https://github.com/apache/spark/commit/524be86158be0a328eecb1634dfd12b6264e0836).
 * This patch merges cleanly.

SparkQA commented:
[Test build #29332 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29332/consoleFull) for   PR 5235 at commit [`524be86`](https://github.com/apache/spark/commit/524be86158be0a328eecb1634dfd12b6264e0836).
 * This patch **passes all tests**.

 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29332/
Test PASSed.

srowen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala:
```diff
@@ -179,6 +179,7 @@ trait ScalaReflection {
       case t if t <:< definitions.ShortTpe => Schema(ShortType, nullable = false)
       case t if t <:< definitions.ByteTpe => Schema(ByteType, nullable = false)
       case t if t <:< definitions.BooleanTpe => Schema(BooleanType, nullable = false)
+      case other => sys.error(s"Schema for type $other is currently not supported")
```
Nits: `case _ =>` ? and rather than raise `RuntimeException`, how about explicitly `IlegalArgumentException` or `UnsupportedOperationException`?

marmbrus replied:
You need the `case other` since we want to print the invalid type.  I'd probably use `AnalysisException` as that is what we throw for the case when there is something wrong with a user's query,

eranation replied:
This is my first pull request so thank you for your patience and feedback!

I used sys.error instead of a more explicit exception for consistency with other similar cases (e.g. in 1.2.1 a non byte array had the same handling using a sys.error

I think that `UnsupportedOperationException` is the most fitting exception if you guys want to go with an explicit one. Or keep it as sys.error to keep it consistent with other error handing in this method. 

What do you prefer?

srowen replied:
(Yeah I corrected my comment right after I made it -- of course `other` is required here.) I don't feel strongly about it; all else equal `RuntimeException` is not ideal, but it's more important to be consistent with similar code than anything.

marmbrus replied:
I would not follow the existing convention of `RuntimeException` as we are currently making an effort to improve errors throughout Spark SQL.  I guess I agree that `UnsupportedOperationException` is probably most accurate here.

eranation committed:
Change from sys.error UnsupportedOperationException

Per code review recommendation, use UnsupportedOperationException (in java.lang so no new imports) as it is more explicit than a general sys.error (which delegates to a RuntimeException)

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29368/
Test FAILed.

eranation commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala:
```diff
@@ -179,6 +179,7 @@ trait ScalaReflection {
       case t if t <:< definitions.ShortTpe => Schema(ShortType, nullable = false)
       case t if t <:< definitions.ByteTpe => Schema(ByteType, nullable = false)
       case t if t <:< definitions.BooleanTpe => Schema(BooleanType, nullable = false)
+      case other => throw new UnsupportedOperationException(s"Schema for type $other is currently not supported")
```
Seems like the DDOS on Github has it's affect on tests...
Tests failed due to

stdout: 
stderr: error:  while accessing https://github.com/apache/spark.git/info/refs
fatal: HTTP request failed

Hope it will be back online soon...

rxin replied:
I will rerun the test, but fyi this will fail the style checker because the line is over 100 chars.

eranation replied:
Thanks @rxin! 
What is the preferred way to split a line then? 

Just starting a new line won't cut it (still at 102 chars)
I can use a temp var, the cost is negligible (we are throwing a runtime exception, it's orders of magnitude more than a redundant assignment)

    case other => 
        val errorMessage = s"Schema for type $other is currently not supported"
        throw new UnsupportedOperationException(errorMessage)

I think this is more readable than this 

    case other => 
        throw new UnsupportedOperationException(s"Schema for type $other is " +
            "currently not supported")

right?

in any case, thanks for your patience and welcoming attitude, I know this is pretty much the lowest priority, but I truly appreciate the fact you guys help out a first timer, this really speaks a lot on the team. Not taken for granted.

eranation commented:
Seems the DDOS has taken its toll on tests. 
Test failed error message:

stderr: error:  while accessing https://github.com/apache/spark.git/info/refs
fatal: HTTP request failed

Hope it will run the tests again one this attack is over :)

rxin commented:
Jenkins, retest this please.

SparkQA commented:
[Test build #29384 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29384/consoleFull) for   PR 5235 at commit [`0c69e9d`](https://github.com/apache/spark/commit/0c69e9d92b83129ee8a668aa810c9ea374d6989f).

SparkQA commented:
[Test build #29384 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29384/consoleFull) for   PR 5235 at commit [`0c69e9d`](https://github.com/apache/spark/commit/0c69e9d92b83129ee8a668aa810c9ea374d6989f).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29384/
Test FAILed.

eranation committed:
Line should be under 100 chars

SparkQA commented:
[Test build #29386 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29386/consoleFull) for   PR 5235 at commit [`af4b1a2`](https://github.com/apache/spark/commit/af4b1a26b58e2e5a25456f6f7f65f4d6b9feff26).

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala:
```diff
@@ -179,6 +179,8 @@ trait ScalaReflection {
       case t if t <:< definitions.ByteTpe => Schema(ByteType, nullable = false)
       case t if t <:< definitions.BooleanTpe => Schema(BooleanType, nullable = false)
+      case other =>
+        throw new UnsupportedOperationException(s"Schema for type $other is not supported")
```
I was actually going to suggest dropping the word "currently". So here you go! :)

eranation replied:
ha :)
awesome

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29386/
Test PASSed.

rxin commented:
Thanks. I'm going to merge this in master & branch-1.3.
